core/embed/rust/src/micropython/runtime.rs

```diff
@@ -1,10 +1,90 @@
-use cstr_core::CStr;
+use core::mem::MaybeUninit;
 
-use super::ffi;
+use crate::error::Error;
 
-pub fn raise_value_error(msg: &'static CStr) -> ! {
+use super::{ffi, obj::Obj};
+
+pub fn raise_exception(err: Error) -> ! {
     unsafe {
-        ffi::mp_raise_ValueError(msg.as_ptr());
+        // SAFETY:
+        // - argument must be an exception instance
+        // (err.to_obj() should return the right thing)
+        ffi::nlr_jump(err.to_obj().as_ptr());
     }
     panic!();
 }
+
+/// Execute `func` while catching MicroPython exceptions. Returns `Ok` in the
+/// successful case, and `Err` with the caught `Obj` in case of a raise.
+pub fn catch_exception<F, T>(mut func: F) -> Result<T, Error>
+where
+    F: FnMut() -> T,
+{
+    // Because MicroPython exceptions use `setjmp` and `longjmp`-like mechanism that
+    // doesn't play too well with Rust, we setup the non-local return pads in C, and
+    // execute `func` through a callback.
+
+    unsafe {
+        // First, we craft a wrapping closure that calls `func`. Because we are generic
+        // over the return type, we cannot pass the returned value over the FFI
+        // boundary, so we assign it explicitly in `wrapper`.
+        let mut result = MaybeUninit::zeroed();
+        let mut wrapper = || {
+            result = MaybeUninit::new(func());
```

Comment:
We can use `MaybeUninit::write()` in 1.55.

Reply:
done in aa7f5c39289700db09d8b5a7fd351b8087061f37